### PR TITLE
feat: add withFileTypes in FileSystem

### PR DESCRIPTION
**Pull request by fu1996**
This pr is from #16780 in the webpack repository. [#16780](https://github.com/webpack/webpack/issues/16780)
The relevant unit tests have been run, and types.d.ts have been automatically generated. Please pr this
This is my first attempt to resolve issues in webpack. If there are any issues that do not meet the specification, I will modify them as soon as possible. 😄

Results of running unit tests locally:
image

### CODE REVIEW

File: types.d.ts
Comment (alexander-akait) [inline types.d.ts:227]: It is possible to pass `undefined` directly here, so `boolean | undefined`

File: types.d.ts
Comment (alexander-akait) [inline types.d.ts:226]: Let's create a new type `BufferEncoding` and move them there and reuse here and  above

File: types.d.ts
Comment (alexander-akait) [inline types.d.ts:126]: `undefined` possible too

**Review by alexander-akait — COMMENTED** [review]
Let's fix some things and we can merge, thank you

File: types.d.ts
Comment (alexander-akait) [reply types.d.ts:226]: Also `buffer` possible too

Comment (alexander-akait) [reply types.d.ts:226]: Just for infromation, here is node types for such method:
```
export function readdir(
        path: PathLike,
        options:
            | {
                  encoding: BufferEncoding | null;
                  withFileTypes?: false | undefined;
              }
            | BufferEncoding
            | undefined
            | null,
        callback: (err: NodeJS.ErrnoException | null, files: string[]) => void
    ): void;
    /**
     * Asynchronous readdir(3) - read a directory.
     * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
     * @param options The encoding (or an object specifying the encoding), used as the encoding of the result. If not provided, `'utf8'` is used.
     */
    export function readdir(
        path: PathLike,
        options:
            | {
                  encoding: 'buffer';
                  withFileTypes?: false | undefined;
              }
            | 'buffer',
        callback: (err: NodeJS.ErrnoException | null, files: Buffer[]) => void
    ): void;
```

**Comment by fu1996** [comment]
> Let's fix some things and we can merge, thank you

Thank you, cr. It's early morning in China, and I will repair it in 10 hours. 😄

File: types.d.ts
Comment (fu1996) [reply types.d.ts:226]: > Let's create a new type `BufferEncoding` and move them there and reuse here and above


"This types.d.ts is automatically generated by running the 'yarn special lint fix' command. I think what I can do is modify the jsDoc syntax in 'Resolver. js' to generate it, but I cannot control the addition of' ascii '|' utf8 '|' utf-8 '|' utf16le '|' ucs2 '|' ucs-2 '|' base64 '|' latin1 '|' binary '|' hex ';"; "Force the use of the BufferEncoding type, even if I manually modify it after running the 'yarn special lint fix' command, but I don't think this is a good method. Do you have any good ideas for controlling the use of the BufferEncoding type instead of its corresponding enumeration value?"?

The following figure shows that after I manually changed to BufferEncoding and executed the 'yarn special lint fix' command, my manual modification was invalid 

image

**Comment by fu1996** [comment]
> Let's fix some things and we can merge, thank you

On the other hand, after I added 'buffer' and undefined and executed the 'yarn special lint fix' command, 'buffer' was successfully automatically generated in the 'types. d.ts' file, but undefined was not generated. For details, see this commit: [commit url](https://github.com/webpack/enhanced-resolve/pull/374/commits/114d565cbf16998cde71268ccf05ac70ed2a83cd) Is this a normal situation? 😊

**Comment by alexander-akait** [comment]
So `undefined` doesn't generate in types, right?

**Comment by fu1996** [comment]
> So `undefined` doesn't generate in types, right?

Yes, undefined cannot be generated by running the command yarn special lint fix. Do you have any good suggestions?

**Comment by TheLarkInn** [comment]
Sorry I'm _JUST_ seeing this! I apologize for the delay @fu1996. This looks awesome thank you so much for the contribution!

**Review by TheLarkInn — APPROVED** [review]

**Comment by TheLarkInn** [comment]
My understanding is that as long as the parameter is _optional_ that the lint-fix will automatically not generated the `undefined` also and that the `optional` parameter satisfies this.

**Comment by TheLarkInn** [comment]
I'll wait on confirmation from @alexander-akait on this before merging but all looks good!

**Review by alexander-akait — APPROVED** [review]

**Comment by TheLarkInn** [comment]
Congrats @fu1996 on your first PR to webpack! 🎉 

多谢!!

**Pull request merged**
